### ISSUE TITLE
[Feature Request]: Support changing cacheVolume of cn and dn after mo is deployed

### ISSUE THREAD

**Issue by aronchanisme** [matrixorigin/matrixone#9335]
### Is there an existing issue for the same feature request?

- [X] I have checked the existing issues.

### Is your feature request related to a problem?

```Markdown
Currently when changing the cacheVolume conf of cn or dn after mo is deployed, by command 
`kubectl edit mo mo -nmo-hn`, cn or dn pod will be deleted and restarted by operator. But due to some reasons, the pod can be deleted but not able to be pulled up again.
```

### Describe the feature you'd like

Support changing cacheVolume of cn and dn after mo is deployed. For example, 
```yaml
tp:
  cacheVolume:
    size: 10Gi
    storageClassName: local-path
```

### Describe implementation you've considered

none

### Documentation, Adoption, Use Case, Migration Strategy

```Markdown
none
```

### Additional information

<img width="1339" alt="企业微信截图_036d8673-3f38-4c83-85dc-26e4bd8fbf19" src="https://user-images.githubusercontent.com/46433151/236805887-629ea835-7cd3-47f5-b437-6beb714b92ff.png">

**Comment by aronchanisme** [comment]
Test whether changing the cache volume and disk cache settings of a cn set will take effect. The result is negative.  @aylei 

operator: `matrixone-operator-0.8.0-alpha.7` with image `matrixorigin/matrixone-operator:0.8.0-alpha.7`
mo: `nightly-6a5db18a`

before changing
cacheVolume: 98Gi
diskCache: 98Gi
<img width="590" alt="image" src="https://github.com/matrixorigin/matrixone/assets/46433151/d391ef3b-5283-4cc8-af10-ac3d0dc1ba59">
<img width="423" alt="image" src="https://github.com/matrixorigin/matrixone/assets/46433151/d9f4fe67-f77e-4389-8252-8dc0c663a52e">
<img width="802" alt="image" src="https://github.com/matrixorigin/matrixone/assets/46433151/34d664b6-8071-43be-b6ba-1c472f45b24f">


after changing
cacheVolume: 50Gi
diskCache: 48Gi

<img width="1186" alt="image" src="https://github.com/matrixorigin/matrixone/assets/46433151/975b98ea-e6f5-4aae-bb2d-8caa699e6263">
<img width="612" alt="image" src="https://github.com/matrixorigin/matrixone/assets/46433151/186042dc-54c7-4d7a-a949-088f70d8f57c">
<img width="928" alt="image" src="https://github.com/matrixorigin/matrixone/assets/46433151/6cf2f57b-29ae-48ae-ab4a-d01a5bd02610">


Seems like changes are accepted but not applied.  PVC is still the previous one.  And diskCache is still the same.

**Comment by aylei** [comment]
> Test whether changing the cache volume and disk cache settings of a cn set will take effect. The result is negative. @aylei
> 
> operator: `matrixone-operator-0.8.0-alpha.7` with image `matrixorigin/matrixone-operator:0.8.0-alpha.7` mo: `nightly-6a5db18a`
> 
> before changing cacheVolume: 98Gi diskCache: 98Gi <img alt="image" width="590" src="https://user-images.githubusercontent.com/46433151/260969654-d391ef3b-5283-4cc8-af10-ac3d0dc1ba59.png"> <img alt="image" width="423" src="https://user-images.githubusercontent.com/46433151/260970038-d9f4fe67-f77e-4389-8252-8dc0c663a52e.png"> <img alt="image" width="802" src="https://user-images.githubusercontent.com/46433151/260970184-34d664b6-8071-43be-b6ba-1c472f45b24f.png">
> 
> after changing cacheVolume: 50Gi diskCache: 48Gi
> 
> <img alt="image" width="1186" src="https://user-images.githubusercontent.com/46433151/260970889-975b98ea-e6f5-4aae-bb2d-8caa699e6263.png"> <img alt="image" width="612" src="https://user-images.githubusercontent.com/46433151/260971097-186042dc-54c7-4d7a-a949-088f70d8f57c.png"> <img alt="image" width="928" src="https://user-images.githubusercontent.com/46433151/260971284-6cf2f57b-29ae-48ae-ab4a-d01a5bd02610.png">
> Seems like changes are accepted but not applied. PVC is still the previous one. And diskCache is still the same.

Right, this feature is not yet supported

**Comment by aylei** [comment]
Discussed with @aronchanisme , postpone to next release but will fix this ASAP. Enabling this feature only requires upgrading mo-operator.

**Comment by dengn** [comment]
@yangj1211 For doc side, this relates to MO deployment in K8s and support for dynamic change.

**Comment by aylei** [comment]
Whether the PV can be resized in-place is determined by the CSI driver, to get better compatibility, modifying cacheVolume size (and type) would recreate CN/TN Pod

**Comment by aressu1985** [comment]
和阿磊沟通，这个会在2.0.1版本解决

**Comment by aylei** [comment]
Test

1. make sure the storageclass used to create mo is set to `allowVolumeExpansion: true`, and the CSI driver support volume expansion:

```
$ kubectl get storageclass standard -oyaml
allowVolumeExpansion: true
```

2. launch an mo cluster with MatrixOneCluster API using the above storageclass
3. increase TN/Log/CN volume size and check the PV size with `kubectl get pv`

Notes:

1. volumeSize can not be decreased:

```
# * spec.volume.size: Invalid value: resource.Quantity{i:resource.int64Amount{value:9663676416, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"9Gi", Format:"BinarySI"}: volume size cannot be decreased
```

2. volume size can be increased
4. increase volume size will update PVC size of Pods in-place

```
  resources:
    requests:
      storage: 11Gi
```

5. if the storageclass used does not support volume expansion, the update will fail with the following error log in mo-operator

```
sync LogSet: admission webhook \"vlogset.kb.io\" denied the request: LogSet.core.matrixorigin.io \"mo\" is invalid: spec.volume.size: Invalid value:
```

6. if the CSI driver does not support volume expansion, the PVC size will be pending on update:

```
Events:
  Type     Reason             Age   From           Message
  ----     ------             ----  ----           -------
  Warning  ExternalExpanding  27s   volume_expand  waiting for an external controller to expand this PVC
```

**Comment by aylei** [comment]
please kindly verify @aronchanisme

**Comment by aronchanisme** [comment]
Confirmed with POSITIVE result, closing this thread.

ENV: TKE cluster

VERSIONS:
 - mo-operator: `nightly-684a75a-2024-11-20`
 - mo: `main_44f4d5b`

TEST_RECORD:
```bash
[root@VM-0-122-tencentos ~]# k get pod -n matrixone -oyaml |grep -i image 
      image: registry.cn-hangzhou.aliyuncs.com/moc-pub/matrixone-operator:nightly-684a75a-2024-11-20
      imagePullPolicy: IfNotPresent
    imagePullSecrets:
      image: registry.cn-hangzhou.aliyuncs.com/moc-pub/matrixone-operator:nightly-684a75a-2024-11-20
      imageID: registry.cn-hangzhou.aliyuncs.com/moc-pub/matrixone-operator@sha256:dd7d06b424f1c708f4b308ca3c0ef87c38718ba951592adf5d370f8857db82ac

[root@VM-0-122-tencentos ~]# k get mo -n mo-cus-reg
NAME               LOG   DN    VERSION        PHASE   AGE
mo-customer-test   3           main_44f4d5b   Ready   4h22m

[root@VM-0-122-tencentos ~]# k get pvc -n mo-cus-reg 
NAME                                       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
mo-data-mo-customer-test-cnset1-cn-4v8q5   Bound    pvc-e55bf696-3fe1-4cb5-9047-67ad2340c369   100Gi      RWO            cbs-hssd       4h21m
mo-data-mo-customer-test-cnset1-cn-dbtl7   Bound    pvc-3619a694-ece7-4bcd-bb30-665b49c3b96f   100Gi      RWO            cbs-hssd       4h21m
mo-data-mo-customer-test-dn-0              Bound    pvc-77432dcb-71c4-4c6d-a735-af269e00824c   50Gi       RWO            cbs-hssd       4h21m
mo-data-mo-customer-test-log-0             Bound    pvc-46d64867-81bb-41fa-bf5c-98dbef6c1101   100Gi      RWO            cbs-hssd       4h22m
mo-data-mo-customer-test-log-1             Bound    pvc-4352ed0e-bc9d-48ae-bf90-08aab04ced44   100Gi      RWO            cbs-hssd       4h22m
mo-data-mo-customer-test-log-2             Bound    pvc-74d481e2-a89c-471b-9245-76aa4183007f   100Gi      RWO            cbs-hssd       4h22m

[root@VM-0-122-tencentos ~]# k get pv |grep customer
pvc-3619a694-ece7-4bcd-bb30-665b49c3b96f   100Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-cnset1-cn-dbtl7                                                          cbs-hssd                   4h21m
pvc-4352ed0e-bc9d-48ae-bf90-08aab04ced44   100Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-log-1                                                                    cbs-hssd                   4h22m
pvc-46d64867-81bb-41fa-bf5c-98dbef6c1101   100Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-log-0                                                                    cbs-hssd                   4h22m
pvc-74d481e2-a89c-471b-9245-76aa4183007f   100Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-log-2                                                                    cbs-hssd                   4h22m
pvc-77432dcb-71c4-4c6d-a735-af269e00824c   50Gi       RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-dn-0                                                                     cbs-hssd                   4h22m
pvc-e55bf696-3fe1-4cb5-9047-67ad2340c369   100Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-cnset1-cn-4v8q5                                                          cbs-hssd                   4h21m

[root@VM-0-122-tencentos ~]# k get storageclass cbs-hssd -oyaml |grep allowVolumeExpansion
allowVolumeExpansion: true

# change CN cacheVolume from 100G to 150G, and TN cacheVolume from 50G to 80G
[root@VM-0-122-tencentos ~]# k edit mo  -n mo-cus-reg mo-customer-test
matrixonecluster.core.matrixorigin.io/mo-customer-test edited

[root@VM-0-122-tencentos ~]# k get pod -n mo-cus-reg -owide
NAME                               READY   STATUS    RESTARTS   AGE     IP               NODE           NOMINATED NODE   READINESS GATES
mo-customer-test-cnset1-cn-4v8q5   1/1     Running   0          4h22m   10.142.182.186   10.129.0.108   <none>           3/3
mo-customer-test-cnset1-cn-dbtl7   1/1     Running   0          4h22m   10.142.182.185   10.129.1.154   <none>           3/3
mo-customer-test-dn-0              1/1     Running   0          4h23m   10.142.182.184   10.129.1.130   <none>           2/2
mo-customer-test-log-0             1/1     Running   0          4h23m   10.142.182.181   10.129.1.120   <none>           2/2
mo-customer-test-log-1             1/1     Running   0          4h23m   10.142.182.182   10.129.1.55    <none>           2/2
mo-customer-test-log-2             1/1     Running   0          4h23m   10.142.182.180   10.129.0.160   <none>           2/2
mo-customer-test-proxy-lb69f       1/1     Running   0          4h23m   10.142.182.183   10.129.1.154   <none>           2/2

# pvc of CN changed to 150G, TN changed to 80G
[root@VM-0-122-tencentos ~]# k get pvc -n mo-cus-reg 
NAME                                       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
# check this 
mo-data-mo-customer-test-cnset1-cn-4v8q5   Bound    pvc-e55bf696-3fe1-4cb5-9047-67ad2340c369   150Gi      RWO            cbs-hssd       4h23m
# check this 
mo-data-mo-customer-test-cnset1-cn-dbtl7   Bound    pvc-3619a694-ece7-4bcd-bb30-665b49c3b96f   100Gi      RWO            cbs-hssd       4h23m
# check this 
mo-data-mo-customer-test-dn-0              Bound    pvc-77432dcb-71c4-4c6d-a735-af269e00824c   80Gi       RWO            cbs-hssd       4h24m
mo-data-mo-customer-test-log-0             Bound    pvc-46d64867-81bb-41fa-bf5c-98dbef6c1101   100Gi      RWO            cbs-hssd       4h24m
mo-data-mo-customer-test-log-1             Bound    pvc-4352ed0e-bc9d-48ae-bf90-08aab04ced44   100Gi      RWO            cbs-hssd       4h24m
mo-data-mo-customer-test-log-2             Bound    pvc-74d481e2-a89c-471b-9245-76aa4183007f   100Gi      RWO            cbs-hssd       4h24m

# pv the same
[root@VM-0-122-tencentos ~]# k get pv |grep customer
# check this 
pvc-3619a694-ece7-4bcd-bb30-665b49c3b96f   150Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-cnset1-cn-dbtl7                                                          cbs-hssd                   4h23m
pvc-4352ed0e-bc9d-48ae-bf90-08aab04ced44   100Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-log-1                                                                    cbs-hssd                   4h24m
pvc-46d64867-81bb-41fa-bf5c-98dbef6c1101   100Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-log-0                                                                    cbs-hssd                   4h24m
pvc-74d481e2-a89c-471b-9245-76aa4183007f   100Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-log-2                                                                    cbs-hssd                   4h24m
# check this 
pvc-77432dcb-71c4-4c6d-a735-af269e00824c   80Gi       RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-dn-0                                                                     cbs-hssd                   4h24m
# check this 
pvc-e55bf696-3fe1-4cb5-9047-67ad2340c369   150Gi      RWO            Delete           Bound    mo-cus-reg/mo-data-mo-customer-test-cnset1-cn-4v8q5                                                          cbs-hssd                   4h23m

# verify pod mounted volume, positive results
[root@VM-0-122-tencentos ~]# k exec -it -n mo-cus-reg mo-customer-test-cnset1-cn-4v8q5 -- bash
root@mo-customer-test-cnset1-cn-4v8q5:/# df -Th 
Filesystem     Type     Size  Used Avail Use% Mounted on
overlay        overlay   50G   38G  9.4G  81% /
tmpfs          tmpfs     64M     0   64M   0% /dev
tmpfs          tmpfs     16G     0   16G   0% /sys/fs/cgroup
/dev/vda1      ext4      50G   38G  9.4G  81% /etc/hosts
shm            tmpfs     64M     0   64M   0% /dev/shm
/dev/vdb       ext4     148G   19G  129G  13% /var/lib/matrixone  # check this 
tmpfs          tmpfs     26G   12K   26G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs          tmpfs     16G     0   16G   0% /proc/acpi
tmpfs          tmpfs     16G     0   16G   0% /proc/scsi
tmpfs          tmpfs     16G     0   16G   0% /sys/firmware
root@mo-customer-test-cnset1-cn-4v8q5:/# exit
exit
[root@VM-0-122-tencentos ~]# k exec -it -n mo-cus-reg mo-customer-test-cnset1-cn-dbtl7 -- bash
root@mo-customer-test-cnset1-cn-dbtl7:/# df -Th 
Filesystem     Type     Size  Used Avail Use% Mounted on
overlay        overlay   50G   35G   13G  74% /
tmpfs          tmpfs     64M     0   64M   0% /dev
tmpfs          tmpfs     16G     0   16G   0% /sys/fs/cgroup
/dev/vda1      ext4      50G   35G   13G  74% /etc/hosts
shm            tmpfs     64M     0   64M   0% /dev/shm
/dev/vdb       ext4     148G   16G  132G  11% /var/lib/matrixone    # check this 
tmpfs          tmpfs     26G   12K   26G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs          tmpfs     16G     0   16G   0% /proc/acpi
tmpfs          tmpfs     16G     0   16G   0% /proc/scsi
tmpfs          tmpfs     16G     0   16G   0% /sys/firmware
root@mo-customer-test-cnset1-cn-dbtl7:/# exit
exit
[root@VM-0-122-tencentos ~]# k exec -it -n mo-cus-reg mo-customer-test-dn-0 -- bash
root@mo-customer-test-dn-0:/# df -Th 
Filesystem     Type     Size  Used Avail Use% Mounted on
overlay        overlay   50G   36G   12G  76% /
tmpfs          tmpfs     64M     0   64M   0% /dev
tmpfs          tmpfs     16G     0   16G   0% /sys/fs/cgroup
/dev/vda1      ext4      50G   36G   12G  76% /etc/hosts
shm            tmpfs     64M     0   64M   0% /dev/shm
/dev/vdb       ext4      79G   41G   38G  53% /var/lib/matrixone     # check this 
tmpfs          tmpfs     20G   12K   20G   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs          tmpfs     16G     0   16G   0% /proc/acpi
tmpfs          tmpfs     16G     0   16G   0% /proc/scsi
tmpfs          tmpfs     16G     0   16G   0% /sys/firmware
root@mo-customer-test-dn-0:/# 
root@mo-customer-test-dn-0:/# exit
exit

# also noted cn and tn pods are not restarted after persistent volume expansion is done
[root@VM-0-122-tencentos ~]# k get pod -n mo-cus-reg -owide
NAME                               READY   STATUS    RESTARTS   AGE     IP               NODE           NOMINATED NODE   READINESS GATES
mo-customer-test-cnset1-cn-4v8q5   1/1     Running   0          4h26m   10.142.182.186   10.129.0.108   <none>           3/3
mo-customer-test-cnset1-cn-dbtl7   1/1     Running   0          4h26m   10.142.182.185   10.129.1.154   <none>           3/3
mo-customer-test-dn-0              1/1     Running   0          4h26m   10.142.182.184   10.129.1.130   <none>           2/2
mo-customer-test-log-0             1/1     Running   0          4h27m   10.142.182.181   10.129.1.120   <none>           2/2
mo-customer-test-log-1             1/1     Running   0          4h27m   10.142.182.182   10.129.1.55    <none>           2/2
mo-customer-test-log-2             1/1     Running   0          4h27m   10.142.182.180   10.129.0.160   <none>           2/2
mo-customer-test-proxy-lb69f       1/1     Running   0          4h26m   10.142.182.183   10.129.1.154   <none> 
```

**Comment by aronchanisme** [comment]
@yangj1211 please kindly help with doc side, thx.  https://github.com/matrixorigin/matrixorigin.io.cn/issues/1167
